woocommerce/woocommerce
# Fix: Product Collection block inside Single Product block was not rendering correctly on frontend.

### Changes proposed in this Pull Request:





Closes #51352

This PR addresses issue where the Product Collection block inside a Single Product block was not rendering correctly on the frontend. Specifically, the first product in the collection was being replaced by the product selected in the parent Single Product block on **Frontend**.

For example, if I add any Product Collection outside of a Single Product block, and first product is "Cap". Now if I add a Single Product block and select "Album" as product. Then If I move Product Collection block inside Single Product block, the first product in the collection is "Album" instead of "Cap" on **Frontend**. It works as expected on **Editor** side. This PR fixes this issue.

**Root Cause**

There is a function called [update_context](https://github.com/woocommerce/woocommerce/blob/8954449e96b48a381d315d62179186e4d52be7bf/plugins/woocommerce/src/Blocks/BlockTypes/SingleProduct.php#L77) in `SingleProduct.php` file. Purpose of this function is to inject the correct post data into the context for each of the Single Product inner blocks. 
- First it gets the product id from the `productId` attribute of the `woocommerce/single-product` block.
- Then it extracts the names of all inner blocks.
- Then it updates the context for each of the inner blocks with following data:
```php
$context['postId']        = $this->product_id; // Product ID extracted from `productId` attribute of the `woocommerce/single-product` block.
$context['singleProduct'] = true;
```

Reason for doing this is mentioned [here](https://github.com/woocommerce/woocommerce/blob/8954449e96b48a381d315d62179186e4d52be7bf/plugins/woocommerce/src/Blocks/BlockTypes/SingleProduct.php#L44-L58). As Gutenberg versions 15.2 and 15.6 are now part of the core of WordPress, I believe this is no longer required. CC: @thealexandrelara As you worked on this previously, could you please confirm if this is still required? 

**What's the solution?**

I updated function [extract_single_product_inner_block_names](https://github.com/woocommerce/woocommerce/blob/8954449e96b48a381d315d62179186e4d52be7bf/plugins/woocommerce/src/Blocks/BlockTypes/SingleProduct.php#L101) in `SingleProduct.php` file to exclude all inner blocks of `woocommerce/product-template` block from the list of inner blocks. This way, the `update_context` function will not override context of inner blocks of `woocommerce/product-template` block. CC: @thealexandrelara I don't think this solution could be problematic for Single Product block, but please correct me if I am wrong. 

CC: @xristos3490 JFYI as I believe you raised something similar in the past.




### How to test the changes in this Pull Request:



1. Create a new post and add a Single Product block.
2. Select a specific product for the Single Product block (e.g., "Album").
3. Add a Product Collection block as a child of the Single Product block.
   - Ensure the first product in the collection differs from the one selected in the Single Product block.
4. Publish the post and view it on the frontend.
5. Verify the following:
   - The first product in the Product Collection is not overridden by the product from the Single Product block.
   - The Product Collection displays the same products as seen in the Editor.
   - The order of products in the collection matches the Editor view.
6. (Step for developers only)
	- Switch to the `trunk` branch and verify that you are able to replicate the issue on `trunk` branch.

### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Fix: Product Collection block inside Single Product block was not rendering correctly on frontend.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
